### PR TITLE
[skrifa] autohint: CJK edges

### PR DESCRIPTION
**Pull request by dfrg**
This was similar enough to latin that changes were folded into the current code with some branches.

### CODE REVIEW

File: skrifa/src/outline/autohint/axis.rs
Comment (rsheeter) [inline skrifa/src/outline/autohint/axis.rs:187]: is there a reason this isn't `next_edge`? Does it mean something different?

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (rsheeter) [inline skrifa/src/outline/autohint/latin/edges.rs:28]: horizontal implies not top-to-bottom makes sense. The group limitation somewhat less so. I suggest a comment.

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (rsheeter) [inline skrifa/src/outline/autohint/latin/edges.rs:147]: Why not `for (ix, segment) in axis.segments.iter().enumerate().filter(no direction)`? - particularly the `.iter().enumerate()`

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (rsheeter) [inline skrifa/src/outline/autohint/latin/edges.rs:148]: Suggest `// Find the first matching edge`

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (rsheeter) [reply skrifa/src/outline/autohint/latin/edges.rs:147]: Would that create unpleasently overlapping borrows? - if so suggest a comment noting as much

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (rsheeter) [inline skrifa/src/outline/autohint/latin/edges.rs:157]: At a glance this also looks possible to do via iterators? - I didn't try so perhaps not; if not perhaps add a comment explaining why.

`if let Some(edge_ix) = axis.edges.iter().enumerate().filter_map(if dist < threshold retain index).find()` type of deal?

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (rsheeter) [inline skrifa/src/outline/autohint/latin/edges.rs:282]: because ...

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (rsheeter) [inline skrifa/src/outline/autohint/latin/edges.rs:317]: I suggest adding some methods to things like `ScaledBlue` and `Edge`. For example, instead of `let is_top = blue.flags & (blue_flags::TOP | blue_flags::LATIN_SUB_TOP) != 0;` we could have `blue.is_top()`, instead of `(edge.fpos as i32 - unscaled_blue.position).abs()` we could have `edge.distance_to(unscaled_blue)` (...or would it need to be specifically unscaled_distance_to?)

This could be deferred to an issue if we don't want to address here. I feel like autohinting would be quite a bit easier to read if widely deployed.

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (rsheeter) [inline skrifa/src/outline/autohint/latin/edges.rs:295]: Optional. I don't have time to try it but finding the best values for an edge smells a lot like reduce. Is it possible it would make sense to write this in terms of reduce? - if it does but we don't want to do it now feel free to defer to an issue.

**Review by rsheeter — APPROVED** [review]

File: skrifa/src/outline/autohint/axis.rs
Comment (dfrg) [reply skrifa/src/outline/autohint/axis.rs]: This returns the next segment in the parent edge. This name was copied from FT but I agree that it's ambiguous. Renamed to `next_in_edge` for clarity.

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (dfrg) [reply skrifa/src/outline/autohint/latin/edges.rs:31]: I'm probably being overly defensive here but the CJK code always passes constant 0 where this value is expected. I added a comment with a link to FT code

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (dfrg) [reply skrifa/src/outline/autohint/latin/edges.rs]: Yes, we make mutable calls into `axis` below which trips up the borrow checker. Added a comment

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (dfrg) [reply skrifa/src/outline/autohint/latin/edges.rs]: Changed to "Try to find an edge that coincides with this segment within the threshold" which better explains what's happening here

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (dfrg) [reply skrifa/src/outline/autohint/latin/edges.rs]: Good catch. Done

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (dfrg) [reply skrifa/src/outline/autohint/latin/edges.rs]: ... FreeType does this. I added a link to the relevant code.

In reality, FT doesn't use any alignment zones in the horizontal direction but the infra exists for it. I believe this is an attempt at better hinting for vertical text but the code is permanently disabled with an `#undef`. After this all lands, I think it's worth considering simplifying our code and removing this altogether. Filed #1140

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (dfrg) [reply skrifa/src/outline/autohint/latin/edges.rs:328]: There's a lot that can be done to clean up the whole autohint module but I prioritized (mostly) matching the FT code in the first pass. I agree that we can do much better but I would like to defer until after we land. Filed #1141

File: skrifa/src/outline/autohint/latin/edges.rs
Comment (dfrg) [reply skrifa/src/outline/autohint/latin/edges.rs:306]: Added to #1141. This seems possible but the interior code would need to be factored out into a function to make the change worthwhile.

**Pull request merged**
